pkg/controller/broadcastjob/broadcastjob_controller.go

```diff
@@ -188,24 +192,50 @@ func (r *ReconcileBroadcastJob) Reconcile(request reconcile.Request) (reconcile.
 	failed := int32(len(failedPods))
 	succeeded := int32(len(succeededPods))
 
-	oldNumConditions := len(job.Status.Conditions)
+	//oldNumConditions := len(job.Status.Conditions)
 	var desired int32
 	desiredNodes, restNodesToRunPod, podsToDelete := getNodesToRunPod(nodes, job, existingNodeToPodMap)
 	desired = int32(len(desiredNodes))
 	klog.Infof("%s/%s has %d/%d nodes remaining to schedule pods", job.Namespace, job.Name, len(restNodesToRunPod), desired)
 	klog.Infof("Before broadcastjob reconcile %s/%s, desired=%d, active=%d, failed=%d", job.Namespace, job.Name, desired, active, failed)
+	job.Status.Active = active
+	job.Status.Failed = failed
+	job.Status.Succeeded = succeeded
+	job.Status.Desired = desired
+
+	if job.Status.Phase == appsv1alpha1.PhaseFailed {
+		return reconcile.Result{RequeueAfter: requeueAfter}, r.updateJobStatus(request, job)
+	}
+
+	if job.Spec.Paused && (job.Status.Phase == appsv1alpha1.PhaseRunning || job.Status.Phase == appsv1alpha1.PhasePaused) {
+		job.Status.Phase = appsv1alpha1.PhasePaused
+		return reconcile.Result{RequeueAfter: requeueAfter}, r.updateJobStatus(request, job)
+	}
+	if job.Spec.Paused == false && job.Status.Phase == appsv1alpha1.PhasePaused {
+		job.Status.Phase = appsv1alpha1.PhaseRunning
+		r.recorder.Event(job, corev1.EventTypeNormal, "Continue", "continue to process job")
+	}
 
 	jobFailed, failureReason, failureMessage := isJobFailed(job, pods)
 	// Job is failed. For keepAlive type, the job will never fail.
 	if jobFailed {
-		// Handle Job failures, delete all active pods
-		failed, active, err = r.deleteJobPods(job, activePods, failed, active)
-		if err != nil {
-			klog.Errorf("failed to deleteJobPods for job %s,", job.Name)
+		switch job.Spec.FailurePolicy {
+		case appsv1alpha1.FailurePolicyPause:
+			r.recorder.Event(job, corev1.EventTypeWarning, "Paused", "job is paused, due to failed pod")
+			job.Spec.Paused = true
+			job.Status.Phase = appsv1alpha1.PhasePaused
+			return reconcile.Result{RequeueAfter: requeueAfter}, r.updateJobStatus(request, job)
+		case appsv1alpha1.FailurePolicyFastFailed:
+			// Handle Job failures, delete all active pods
+			failed, active, err = r.deleteJobPods(job, activePods, failed, active)
+			if err != nil {
+				klog.Errorf("failed to deleteJobPods for job %s,", job.Name)
+			}
+			requeueAfter = finishJob(job, appsv1alpha1.JobFailed, failureMessage)
+			r.recorder.Event(job, corev1.EventTypeWarning, failureReason,
+				fmt.Sprintf("%s: %d pods succeeded, %d pods failed", failureMessage, succeeded, failed))
+		case appsv1alpha1.FailurePolicyContinue:
```

Comment:
Job can fail due to pod failure or hitting ActiveDeadline. Should we continue in the latter case?

Also, it seems that the Continue logic does not work, the reconcilePods method is not called in this code path.